Fatras/include/ActsFatras/Digitization/UncorrelatedHitSmearer.hpp

```diff
@@ -0,0 +1,127 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/EventData/ParameterSet.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/Surfaces/SurfaceError.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+#include "Acts/Utilities/ParameterDefinitions.hpp"
+#include "Acts/Utilities/Result.hpp"
+#include "ActsFatras/Digitization/detail/ParametersSmearer.hpp"
+#include "ActsFatras/EventData/Hit.hpp"
+
+namespace ActsFatras {
+
+/// Smearing functions definition:
+/// - it takes the unsmeared parameter
+/// - it returns the smeared parameter and a covariance
+using SmearFunction =
+    std::function<Acts::Result<std::pair<double, double>>(double)>;
+
+/// Parameter smearer for fast digitisation that produces
+///
+///
+/// The logic of this smearer is the following:
+/// - Input is a single simulated ActsFatras::Hit which is wrapped into
+/// - It returns smeared parameter vector and a covariance matrix
+///
+/// @note This smearer only supports uncorrelated smearing of parameters
+///
+class UncorrelatedHitSmearer {
+ public:
+  /// Generic implementation of a smearing meathod for bound parameters
+  ///
+  /// @tparam kParameters parameter pack describing the parameters to smear
+  ///
+  /// @param gctx The Geometry context (alignment, etc.)
+  /// @param hit The Fatras hit to be smeared (will be source linked)
+  /// @param sf The Surface to which this hit is attached
+  /// @param sFunctions The smearing functions that are applied
+  ///
+  /// @return Smeared bound parameter vector and covariance matrix
+  template <Acts::BoundIndices... kParameters>
+  Acts::Result<Acts::ParameterSet<Acts::BoundIndices, kParameters...>>
+  smearedParameterSet(const Acts::GeometryContext& gctx, const Hit& hit,
+                      const Acts::Surface& sf,
+                      const std::array<SmearFunction, sizeof...(kParameters)>&
+                          sFunctions) const {
+    using ParSet = Acts::ParameterSet<Acts::BoundIndices, kParameters...>;
+    using Result = Acts::Result<ParSet>;
+    using ParametersSmearer =
+        detail::ParametersSmearer<Acts::BoundIndices, kParameters...>;
+
+    auto dir = hit.unitDirection();
+    auto gltResult = sf.globalToLocal(gctx, hit.position(), dir);
+    if (not gltResult.ok()) {
+      return Result(Acts::SurfaceError::GlobalPositionNotOnSurface);
+    }
+    const auto& lPosition = gltResult.value();
+
+    typename ParSet::FullParametersVector fParameters;
+    fParameters.setZero();
+    fParameters.template segment<2>(0) = lPosition;
+    fParameters[Acts::eBoundPhi] = Acts::VectorHelpers::phi(dir);
```

Comment:
Maybe I misunderstand something, but I would expect `if (ParSet::contains<eBoundPhi>())` to be called here, no?

Reply:
As I said, we can put protect this with if statements if necessary, in my recent work I have seen that if branching is sometimes more CPU costly than just performing the operation and not introduce branch predictions.

Reply:
(But you are right, this is the place where we could check).

Reply:
Ok, I see.